rockcraft/extensions/flask.py

```diff
@@ -0,0 +1,105 @@
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""An experimental extension for the Flask framework."""
+
+import os
+from typing import Any, Dict, Optional, Tuple
+
+from overrides import override
+
+from .extension import Extension
+
+
+class Flask(Extension):
+    """An extension for constructing Python applications based on the Flask framework."""
+
+    @staticmethod
+    @override
+    def get_supported_bases() -> Tuple[str, ...]:
+        """Return supported bases."""
+        return "bare", "ubuntu:22.04"
+
+    @staticmethod
+    @override
+    def is_experimental(base: Optional[str]) -> bool:
+        """Check if the extension is in an experimental state."""
+        return True
+
+    @override
+    def get_root_snippet(self) -> Dict[str, Any]:
+        """Fill in some default root components for Flask.
+
+        Default values:
+          - run_user: _daemon_
+          - build-base: ubuntu:22.04 (only if user specify bare without a build-base)
+          - platform: amd64
+        """
+        snippet: Dict[str, Any] = {}
+        if "run_user" not in self.yaml_data:
+            snippet["run_user"] = "_daemon_"
+        if (
+            "build-base" not in self.yaml_data
+            and self.yaml_data.get("base", "bare") == "bare"
+        ):
+            snippet["build-base"] = "ubuntu:22.04"
+        if "platforms" not in self.yaml_data:
+            snippet["platforms"] = {"amd64": {}}
+        return snippet
+
+    @override
+    def get_part_snippet(self) -> Dict[str, Any]:
+        """Return the part snippet to apply to existing parts."""
+        return {}
+
+    @override
+    def get_parts_snippet(self) -> Dict[str, Any]:
+        """Create necessary parts to facilitate the flask application.
+
+        Parts added:
+            - flask/dependencies: install Python dependencies
+            - flask/install-app: copy the flask project into the OCI image
+        """
+        if not (self.project_root / "requirements.txt").exists():
+            raise ValueError(
+                "missing requirements.txt file, "
+                "flask extension requires this file with flask specified as a dependency"
+            )
+        ignores = [".git", "node_modules"]
+        source_files = [
+            f
+            for f in os.listdir(self.project_root)
+            if f not in ignores and not f.endswith(".rock")
+        ]
+        renaming_map = {f: os.path.join("srv/flask/app", f) for f in source_files}
```

Comment:
how would one know this is the place where things need to be put?

Reply:
We need to choose a location that is agreed upon by both the charm and the rock for it to function properly. The path `/srv/flask/app` is the outcome of this agreement.